[ATen] Remove explicit casting of complex nansum during accumulation

PaulZhang12 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #165187
* #165178
* #165055
* #164790
* __->__ #165494

https://github.com/pytorch/pytorch/pull/164790 modifies aten to perform a different reduction order intra warp. However, this change exposed a large difference in a sum for complex32. Namely the case:

```
import torch

a = torch.tensor([[ 4.82031250+7.34765625j,
           -3.37109375-1.9501953125j],

         [ 3.7832031250-2.43359375j,
           -6.07812500+5.32812500j]], dtype=torch.complex32, device='cuda:0')

sum_out = torch.sum(a)
nansum_out = torch.nansum(a)
torch.testing.assert_close(
    sum_out,
    nansum_out,
    rtol=0,
    atol=0,
)
```

Here, the result of `sum` and `nansum` differed significantly by 1e-2. Further investigation showed that the explicit casting of b back to `arg_t` from `scalar_t` was the root cause. `arg_t` is the dtype of the accumulator, ComplexFloat, and `scalar_t` of the input dtype, ComplexHalf. When we cast in the reduction to the accumulator order, that means the input is still of ComplexHalf, which loses precision as it can store intermediate values.

PaulZhang12 committed:
Remove explicit casting of complex nansum during accumulation

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -392,8 +392,8 @@ struct NormTwoOps {
 
 template <typename acc_t, typename data_t>
 struct NanSumOps {
-  inline C10_DEVICE acc_t reduce(acc_t a, data_t b, int64_t /*idx*/) const {
```
this is actually fine, `reduce` is called only to add a single number `b` which would be representable in `data_t` to accumulator

ngimel commented on aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:
```diff
@@ -48,6 +48,7 @@ struct sum_functor<c10::complex<at::Half>> {
       return a + b;
     }
     );
+    std::cout << "Jiterator" << std::endl;
```
stray print

PaulZhang12 committed:
Update on "[ATen] Remove explicit casting of complex nansum during accumulation"


https://github.com/pytorch/pytorch/pull/164790 modifies aten to perform a different reduction order intra warp. However, this change exposed a large difference in a sum for complex32. Namely the case:

```
import torch

a = torch.tensor([[ 4.82031250+7.34765625j,
           -3.37109375-1.9501953125j],

         [ 3.7832031250-2.43359375j,
           -6.07812500+5.32812500j]], dtype=torch.complex32, device='cuda:0')

sum_out = torch.sum(a)
nansum_out = torch.nansum(a)
torch.testing.assert_close(
    sum_out,
    nansum_out,
    rtol=0,
    atol=0,
)
```

Here, the result of `sum` and `nansum` differed significantly by 1e-2. Further investigation showed that the explicit casting of b back to `arg_t` from `scalar_t` was the root cause. `arg_t` is the dtype of the accumulator, ComplexFloat, and `scalar_t` of the input dtype, ComplexHalf. When we cast in the reduction to the accumulator order, that means the input is still of ComplexHalf, which loses precision as it can store intermediate values.



[ghstack-poisoned]
